### PR TITLE
Disable optimization on gcc <7

### PR DESCRIPTION
**Pull request by matthiakl**
Re #5100
In a local environment this fixed the ASAN testsuites errors. Let's see how it goes.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -189,9 +189,14 @@ if(CMAKE_BUILD_TYPE STREQUAL "Debug")
...
 elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
-  if(("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU") AND (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 6.2))
-    message(STATUS "GCC >6.2.0 breaks -03, setting -02")
-    set(WL_OPTIMIZE_FLAGS "-O2")
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 7.0)
+      message(STATUS "GCC <7 produce stack-buffer-overflow with -O2, setting -00")
+      set(WL_OPTIMIZE_FLAGS "-O0")
+    else()
+      message(STATUS "GCC >6.2.0 breaks -03, setting -02")
```
Comment (tothxa) [inline CMakeLists.txt]:
```suggestion
      message(STATUS "GCC <7 produce stack-buffer-overflow with -O2, setting -O0")
      set(WL_OPTIMIZE_FLAGS "-O0")
    else()
      message(STATUS "GCC >6.2.0 breaks -O3, setting -O2")
```

**Comment by Noordfrees** [comment]
I had thought about trying this as well but I'm not sure it's the right approach. It's only a small number functions which are reproducibly affected so I'd prefer to fix only those few code pieces instead of completely disabling optimization globally. Did you measure how much of a difference this makes in execution time and binary size? If both is negligible it's probably fine but if there's a difference we should use this only as a last resort IMHO.

Edit: Also it always worked fine without ASan even if the buffer overflow might have gone unnoticed so turning optimization off should only be considered when ASan is enabled.

**Comment by matthiakl** [comment]
I did a time measurement with a subset of the testsuite:
`time ./regression_test.py -b build/gcc6/src/widelands -r port`
 - -O0: (50M, 36M stripped)
 ```
Ran 20 tests in 703.711s
real  11m43.777s
user  9m45.796s
sys   0m31.319s
 ```

- -O2: (31M, 29M stripped)
```
Ran 20 tests in 626.817s
real  10m26.884s
user  8m23.952s
sys   0m27.724s
 ```
The size looks ok, but the time difference is not the best.

**Comment by tothxa** [comment]
Did you consider -O1?

**Comment by matthiakl** [comment]
> Did you consider -O1?

Yes, -O1 produces the same errors. From there I tried to disable single optimizations with -fno-xxx, but without success.
I also considered -O0 with some explicit optimizations, but they did not reduce the size significantly.

**Comment by Noordfrees** [comment]
GCC also supports a function attribute `optimize(level)` which might allow us to fix only those functions where optimization actually causes a bug. Would it be possible to write something like
```c++
#if ((__GNUC__ * 100) + __GNUC_MINOR__) < 700
#define NO_OPTIMIZATION __attribute__ ((optimize(0)))
#else 
#define NO_OPTIMIZATION
#endif
```
```c++
void dangerous_function() NO_OPTIMIZATION {
	...
}
```
and apply this macro to a small range of functions until the testsuite is happy

**Comment by matthiakl** [comment]
I also tried this. It started promising, but I quickly ran into errors with stack traces starting from `boost::format`. 
Should I close this PR then and try again, when #5118 is merged?

**Pull request closed without merging**